java/java-impl-inspections/src/com/intellij/codeInspection/reflectiveAccess/JavaLangReflectHandleInvocationChecker.java

```diff
@@ -251,6 +255,32 @@ else if (invokeParent instanceof PsiVariable) {
 
       if (isCallToMethod(typeDefinitionCall, JAVA_LANG_INVOKE_METHOD_TYPE, METHOD_TYPE)) {
         final PsiExpression[] arguments = typeDefinitionCall.getArgumentList().getExpressions();
+        if (arguments.length == 2) {
+          final PsiExpression returnType = findDefinition(arguments[0]);
+          final PsiExpression secondArgument = findDefinition(arguments[1]);
+          final List<PsiExpression> components = getArrayOrListComponents(secondArgument);
+          if (components != null) {
+            // Call to MethodType.methodType(Class<?>, List<Class<?>>) or MethodType.methodType(Class<?>, Class<?>[])
+            final List<PsiExpression> signature = ContainerUtil.prepend(components, returnType);
+            return ContainerUtil.map(signature, parameter -> (() -> getReflectiveType(parameter)));
+          }
+          if (secondArgument instanceof PsiMethodCallExpression) {
+            final PsiMethodCallExpression maybeNestedMethodTypeCall = (PsiMethodCallExpression)secondArgument;
+            if (isCallToMethod(maybeNestedMethodTypeCall, JAVA_LANG_INVOKE_METHOD_TYPE, METHOD_TYPE)) {
+              // Call to MethodType.methodType(Class<?>, MethodType)
+              final List<Supplier<ReflectiveType>> nestedSignature =
+                ourGuard.doPreventingRecursion(methodTypeExpression, false, () -> getLazyMethodSignature(maybeNestedMethodTypeCall));
```

Comment:
RecursionGuard is mainly intended to catch bugs. Here, we can probably limit recursion depth by ourselves, like passing it as a parameter to `composeMethodSignature`? It could be even outer/inner boolean flag, as I don't think we need more than two levels. Also, I'm not sure whether it's a very useful thing to support. If we created a method type right here, why would we rebind the return type? Is it useful?

Reply:
> RecursionGuard is mainly intended to catch bugs. Here, we can probably limit recursion depth by ourselves, like passing it as a parameter to composeMethodSignature? It could be even outer/inner boolean flag, as I don't think we need more than two levels.

1. I have added an explicit `boolean` flag to control recursion. Only `composeMethodSignatureFromReturnTypeAndMethodType` calls `composeMethodSignature` recursively and it does explicitly disallowing further recursion
2. To resolve the innermost `MethodType` in such cases I use the iterative `findInnermostMethodType` method instead of recursion as in the previous version of the PR

> Also, I'm not sure whether it's a very useful thing to support. If we created a method type right here, why would we rebind the return type? Is it useful?

To be honest, I'm not sure either. I've looked up the usages of the `methodType(Class, MethodType)` overload quickly in other projects: JRuby, Jython, Groovy, and could not find one

But here is a potential, artificial example where the inspection could be useful:

```kotlin
class Foo {
    fun bar(list: List<String>): String = "abc"
    fun bar(list: List<Int>): Int = 12
}
```

```java
public class Main {
    private static final MethodHandle BAR_STRING_METHOD_HANDLE;
    private static final MethodHandle BAR_INT_METHOD_HANDLE;


    static {
        try {
            MethodHandles.Lookup lookup = MethodHandles.lookup();
            MethodType stringListMethodType = methodType(String.class, List.class);
            BAR_STRING_METHOD_HANDLE = lookup.findVirtual(Foo.class, "bar", stringListMethodType);
            BAR_INT_METHOD_HANDLE = lookup.findVirtual(Foo.class, "bar", methodType(int.class, stringListMethodType));
        } catch (Throwable t) {
            throw new IllegalStateException(t);
        }
    }

    public void test(Foo foo) throws Throwable {
        String s = (String) BAR_STRING_METHOD_HANDLE.invokeExact(foo, List.of());
        System.out.println("s = " + s);
        long i = (long) BAR_INT_METHOD_HANDLE.invokeExact(foo, Set.of());
        //        ^^ warning                                   ^^ warning
        System.out.println("i = " + i);
    }

    public static void main(String[] args) throws Throwable {
        new Main().test(new Foo());
    }
}
```